compiler/src/iree/compiler/Dialect/HAL/IR/test/device_ops.mlir

```diff
@@ -202,3 +202,21 @@ util.func public @device_queue_flush(
       affinity(%affinity)
   util.return
 }
+
+// -----
+
+// CHECK-LABEL: @device_memoize
+util.func public @device_memoize(
+    // CHECK-SAME: (%[[DEVICE:.+]]: !hal.device, %[[AFFINITY:.+]]: i64)
+    %device: !hal.device, %affinity: i64) -> !hal.command_buffer {
+  // CHECK: hal.device.memoize<%[[DEVICE]] : !hal.device> affinity(%[[AFFINITY]])
+  %result = hal.device.memoize<%device : !hal.device> affinity(%affinity) -> !hal.command_buffer {
+    // CHECK-NEXT: hal.command_buffer.create
+    %cmd = hal.command_buffer.create device(%device : !hal.device)
+                                       mode(OneShot)
+                                 categories("Transfer|Dispatch") : !hal.command_buffer
+    // CHECK-NEXT: hal.return
+    hal.return %cmd : !hal.command_buffer
+  }
```

Comment:
Explicitly only to start - we could start to use them elsewhere now that they exist, though, such as by replacing the MemoizeDeviceQueries pass by just finding device queries and fusing them into memoize regions or creating memoize regions. MaterializeResourceCaches may also be able to use memoize regions instead of having its special per-device logic but changing to just identify the lookup pseudo ops and replacing them with memoize ops that do the loading in-place (though that would rely on the deduplication being improved in the outlining pass here).